nanocoap/nanocoap.c

```diff
@@ -369,3 +377,65 @@ ssize_t coap_well_known_core_default_handler(coap_pkt_t* pkt, uint8_t *buf, \
 
     return coap_build_reply(pkt, COAP_CODE_205, buf, len, payload_len);
 }
+
+size_t get_lenval(uint8_t *buf, uint16_t *val)
+{
+    size_t len = 0;
+
+    if (*val == 13) {
+        *val += *buf;
+        len = 1;
+    }
+    else if (*val == 14) {
+        memcpy(val, buf, 2);
+        *val = htons(*val) + 269;
+        len = 2;
+    }
+
+    return len;
+}
+
+int parse_opt(uint8_t *optpos, coap_opt_t *opt)
+{
+    opt->val = optpos + 1;
+    opt->delta = ((*optpos & 0xf0) >> 4);
+    opt->len =  (*optpos & 0x0f);
+
+    /* make sure delta and len raw values are valid */
+    if ((opt->delta == 15) || (opt->len == 15)) {
+        return -1;
+    }
+
+    opt->val += get_lenval(opt->val, &opt->delta);
+    opt->val += get_lenval(opt->val, &opt->len);
+
+    return (opt->val - optpos) + opt->len;
+}
+
+uint8_t *coap_find_opt(coap_pkt_t *pkt, uint8_t *bufpos,
+                       coap_opt_t *opt, uint16_t optnum)
+{
+    assert(opt);
+
+    /* check if we reached the end of options */
+    if (!bufpos || (*bufpos == 0xff) || (bufpos == pkt->payload)) {
+        return NULL;
+    }
+
+    uint16_t delta = 0;
+
+    do {
```

Comment:
Can we add packet length checks here? Seems like a bogus header leads to crashes here.